scipy-stubs/fft/_realtransforms.pyi

```diff
@@ -1,82 +1,84 @@
-from scipy._typing import Untyped
+from numpy import float64, generic
+from numpy.typing import NDArray
+from scipy._typing import DCTType, NormalizationMode, Untyped
 
 def dctn(
     x: Untyped,
-    type: int = 2,
+    type: DCTType = 2,
     s: Untyped | None = None,
     axes: Untyped | None = None,
-    norm: Untyped | None = None,
+    norm: NormalizationMode | None = None,
     overwrite_x: bool = False,
     workers: Untyped | None = None,
     *,
     orthogonalize: Untyped | None = None,
 ) -> Untyped: ...
 def idctn(
     x: Untyped,
-    type: int = 2,
+    type: DCTType = 2,
     s: Untyped | None = None,
     axes: Untyped | None = None,
-    norm: Untyped | None = None,
+    norm: NormalizationMode | None = None,
     overwrite_x: bool = False,
     workers: Untyped | None = None,
     orthogonalize: Untyped | None = None,
 ) -> Untyped: ...
 def dstn(
     x: Untyped,
-    type: int = 2,
+    type: DCTType = 2,
     s: Untyped | None = None,
     axes: Untyped | None = None,
-    norm: Untyped | None = None,
+    norm: NormalizationMode | None = None,
     overwrite_x: bool = False,
     workers: Untyped | None = None,
     orthogonalize: Untyped | None = None,
 ) -> Untyped: ...
 def idstn(
     x: Untyped,
-    type: int = 2,
+    type: DCTType = 2,
     s: Untyped | None = None,
     axes: Untyped | None = None,
-    norm: Untyped | None = None,
+    norm: NormalizationMode | None = None,
     overwrite_x: bool = False,
     workers: Untyped | None = None,
     orthogonalize: Untyped | None = None,
 ) -> Untyped: ...
 def dct(
-    x: Untyped,
-    type: int = 2,
-    n: Untyped | None = None,
+    x: NDArray[generic],
+    type: DCTType = 2,
+    n: int | None = None,
     axis: int = -1,
-    norm: Untyped | None = None,
+    norm: NormalizationMode | None = None,
     overwrite_x: bool = False,
-    workers: Untyped | None = None,
-    orthogonalize: Untyped | None = None,
-) -> Untyped: ...
+    workers: int | None = None,
+    orthogonalize: bool | None = None,
+) -> NDArray[float64]: ...
```

Comment:
If the input is an array of e.g. `float32`, then the output will also be `float32`. 
In the same way, the output could also be complex.
But if the input has a dtype of e.g. `np.int64`, `np.uint8`, or `np.bool_`, then the output will indeed always be of `np.float64`.

I think the cleanest way to go about this is with 2 or 3 overloads, one for integer array-like input, and the other(s) for floating- and complex- array-likes.

Reply:
As far as I'm concerned, it's also fine if you don't use overloads and `TypeVar`s, and change the return type to `npt.NDArray[np.inexact[Any]]` instead. Even though it's not as "tight" as it could be, it's a lot easier to implement. And all things considered, it'd still be an improvement :)

The underlying thought here is that both `np.floating` and `np.complexfloating` have `np.inexact` as their supertype, so the return type is always correct.

Reply:
I don't mind using overloads, it's just that I'm not sure what's all the numpy dtypes that I need to cover ^^"

> The underlying thought here is that both `np.floating` and `np.complexfloating` have `np.inexact` as their supertype, so the return type is always correct.

There's a catch for return types: This will actually introduce false-positives. For example, I could no longer pass the result of `fft.dct` to `np.median`:

```py
import cv2
import numpy as np
from cv2.typing import MatLike
from scipy import fft

def __cv2_phash(image: MatLike, hash_size: int = 8, highfreq_factor: int = 4):
    img_size = hash_size * highfreq_factor
    image = cv2.cvtColor(image, cv2.COLOR_BGRA2GRAY)
    image = cv2.resize(image, (img_size, img_size), interpolation=cv2.INTER_AREA)
    dct = fft.dct(fft.dct(image, axis=0), axis=1)
    dct_low_frequency = dct[:hash_size, :hash_size]
    median = np.median(dct_low_frequency)
    return dct_low_frequency > median
```
pyright:
```
error: No overloads for "median" match the provided arguments (reportCallIssue)
error: Argument of type "ndarray[Any, dtype[inexact[Incomplete]]]" cannot be assigned to parameter "a" of type "_ArrayLikeFloat_co | _ArrayLikeComplex_co | _ArrayLikeTD64_co | _ArrayLikeObject_co" in function "median"        
    Type "ndarray[Any, dtype[inexact[Incomplete]]]" is not assignable to type "_ArrayLikeFloat_co | _ArrayLikeComplex_co | _ArrayLikeTD64_co | _ArrayLikeObject_co"
      "ndarray[Any, dtype[inexact[Incomplete]]]" is incompatible with protocol "_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]"
        "__array__" is an incompatible type
          No overloaded function matches type "() -> ndarray[Any, _DType_co@_SupportsArray]"
      "ndarray[Any, dtype[inexact[Incomplete]]]" is incompatible with protocol "_NestedSequence[_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]]"
        "__reversed__" is not present
        "count" is not present
        "index" is not present
    ... (reportArgumentType)
```
mypy:
```
error: Argument 1 to "median" has incompatible type "ndarray[Any, dtype[inexact[Any]]]"; expected "_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]] | _NestedSequence[_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]] | bool | int | float | _NestedSequence[bool | int | float]"  [arg-type]
```

And this isn't even a case that's fixable with an `isinstance` check.

(very similar to this are issues with union return types: https://github.com/python/mypy/issues/1693, both of which could be worked around with https://github.com/python/typing/issues/566 when overloads aren't an option)

Reply:
That's a very good point. I'll definitely keep that in mind when writing/reviewing similar `floating` vs `complexfloating` situations from now on 👌🏻.

That `AnyOf` proposal would indeed be very welcome here. Sometimes you can work around it by using e.g. `T: (str, date)`, but that's limited to within the scope of a function/class/type-alias, and would in this case upcast `datetime` to `date`.
So I really like that idea; thanks for sharing.